feat: add docs on creating event triggers (#40239)
supabase/supabase: apps/docs/content/guides/database/postgres/event-triggers.mdx

---
id: 'postgres-event-triggers'
title: 'Event Triggers'
description: 'Automatically execute SQL on database events.'
subtitle: 'Automatically execute SQL on database events.'
---

In Postgres, an [event trigger](https://www.postgresql.org/docs/current/event-triggers.html) is very similar to a [trigger](/docs/guides/database/postgres/triggers), except that it is triggered by database level events (and is usually reserved for [superusers](/docs/guides/database/postgres/roles-superuser))

With our `Supautils` extension (installed automatically for all Supabase projects), the `postgres` user has the ability to create and manage event triggers.

Some use cases for event triggers ares:

- Capturing Data Definition Language (DDL) changes - these are changes to your database schema (though the [pgAudit](/docs/guides/database/extensions/pgaudit) extension provides a more complete solution)
- Monitoring/preventing actions - such as preventing tables from being dropped in Production or enforcing RLS on all new tables

> In this guide, we will go through an example event trigger that prevents accidental dropping of a table

Reviewer: ```suggestion
This guide covers an example event trigger that prevents the accidental dropping of a table
```

## Creating an event trigger

Event triggers can only be created by the `postgres` user so make sure you are authenticated as them. As with triggers, event triggers consist of 2 parts

1. A [Function](/docs/guides/database/functions) which will be executed
2. The actual Event Trigger object, with parameters around when the trigger should be run.

An example of a trigger is:

```sql
CREATE EVENT TRIGGER dont_drop_trigger
ON sql_drop
EXECUTE FUNCTION dont_drop_function();
```

## Trigger functions

A trigger function is a user-defined [Function](/docs/guides/database/functions) that Postgres executes when the trigger is fired.

### Example trigger function

> This example protects any table from being dropped. It can be overridden by temporarily disabling the event trigger: `ALTER EVENT TRIGGER dont_drop_trigger DISABLE;`

Reviewer: ```suggestion
This example protects any table from being dropped. You can override it by temporarily disabling the event trigger: `ALTER EVENT TRIGGER dont_drop_trigger DISABLE;`
```
